dbms/src/Storages/DeltaMerge/ColumnFile/ColumnFileBig.cpp

```diff
@@ -94,6 +99,59 @@ ColumnFilePersistedPtr ColumnFileBig::deserializeMetadata(const DMContext & cont
     return std::shared_ptr<ColumnFileBig>(dp_file);
 }
 
+ColumnFilePersistedPtr ColumnFileBig::createFromCheckpoint(DMContext & context, //
+                                                           const RowKeyRange & target_range,
+                                                           ReadBuffer & buf,
+                                                           UniversalPageStoragePtr temp_ps,
+                                                           TableID ns_id,
+                                                           WriteBatches & wbs)
+{
+    UInt64 file_page_id;
+    size_t valid_rows, valid_bytes;
+
+    readIntBinary(file_page_id, buf);
+    readIntBinary(valid_rows, buf);
+    readIntBinary(valid_bytes, buf);
+
+    // get target dtfile s3 key
+    auto remote_page_id = UniversalPageIdFormat::toFullPageId(UniversalPageIdFormat::toFullPrefix(StorageType::Data, ns_id), file_page_id);
+    auto remote_data_location = temp_ps->getCheckpointLocation(remote_page_id);
+    auto remote_file_id = temp_ps->getNormalPageId(remote_page_id);
+    auto file_id = UniversalPageIdFormat::getU64ID(remote_file_id);
+
+    const auto & lock_key_view = S3::S3FilenameView::fromKey(*(remote_data_location->data_file_id));
```

Comment:
The dtfile on S3 needs to be locked before put into local metadata(PS). But I can't find where it is locked.

Reply:
It is done inside UniPS::write
https://github.com/pingcap/tiflash/blob/5d9741fa18a58c93cbb2ff7f5befd2569f699421/dbms/src/Storages/Page/V3/Universal/UniversalPageStorage.cpp#L73-L98